File: implementations/csharp/string-search/v8-knuth-morris-pratt-d7l8/Program.cs
using System;
using System.Collections.Generic;
using System.Linq;

public static class KmpSearch
{
  public static int[] Search(string text, string pattern)
  {
    if (pattern.Length == 0) return Array.Empty<int>();
    var lps = new int[pattern.Length];
    for (int i = 1, len = 0; i < pattern.Length;)
    {
      if (pattern[i] == pattern[len]) lps[i++] = ++len;
      else if (len != 0) len = lps[len - 1];
      else lps[i++] = 0;
    }
    var matches = new List<int>();
    for (int i = 0, j = 0; i < text.Length;)
    {
      if (text[i] == pattern[j])
      {
        i++;
        j++;
        if (j == pattern.Length)
        {
          matches.Add(i - j);
          j = lps[j - 1];
        }
      }
      else if (j != 0) j = lps[j - 1];
      else i++;
    }
    return matches.ToArray();
  }

  public static void Main()
  {
    Expect("ababcabcabababd", "ababd", new[] {10});
    Expect("aaaaa", "aa", new[] {0, 1, 2, 3});
    Expect("abcdef", "gh", Array.Empty<int>());
    Expect("abc", "abc", new[] {0});
    Console.WriteLine("csharp kmp ok");
  }

  static void Expect(string text, string pattern, int[] expected)
  {
    if (!Search(text, pattern).SequenceEqual(expected)) throw new Exception(pattern);
  }
}

File: implementations/fsharp/string-search/v8-knuth-morris-pratt-d7l8/Program.fs
let search (text: string) (pattern: string) =
    if pattern.Length = 0 then
        [||]
    else
        let lps = Array.zeroCreate<int> pattern.Length
        let mutable len = 0
        let mutable i = 1
        while i < pattern.Length do
            if pattern[i] = pattern[len] then
                len <- len + 1
                lps[i] <- len
                i <- i + 1
            elif len <> 0 then
                len <- lps[len - 1]
            else
                lps[i] <- 0
                i <- i + 1
        let matches = ResizeArray<int>()
        let mutable ti = 0
        let mutable pj = 0
        while ti < text.Length do
            if text[ti] = pattern[pj] then
                ti <- ti + 1
                pj <- pj + 1
                if pj = pattern.Length then
                    matches.Add(ti - pj)
                    pj <- lps[pj - 1]
            elif pj <> 0 then
                pj <- lps[pj - 1]
            else
                ti <- ti + 1
        matches.ToArray()

let expect text pattern expected =
    let actual = search text pattern
    if actual <> expected then failwithf "%s failed" pattern

[<EntryPoint>]
let main _ =
    expect "ababcabcabababd" "ababd" [|10|]
    expect "aaaaa" "aa" [|0; 1; 2; 3|]
    expect "abcdef" "gh" [||]
    expect "abc" "abc" [|0|]
    printfn "fsharp kmp ok"
    0
